File: core/src/main/java/com/tek/core/TekModuleConfiguration.java
package com.tek.core;

import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.env.ConfigurableEnvironment;
import org.springframework.util.ClassUtils;

import javax.annotation.PostConstruct;
import javax.naming.ConfigurationException;

/**
 * Template class to create a Tek module configuration.
 *
 * @author MarcoPagan
 */
@Slf4j
public abstract class TekModuleConfiguration {

    @Autowired
    protected ConfigurableEnvironment environment;

    private final Class<?> configuration;

    protected TekModuleConfiguration(Class<?> configuration) {
        this.configuration = configuration;
    }

    public abstract void checkModuleConfiguration() throws ConfigurationException;

    @PostConstruct
    @SneakyThrows
    private void postConstruct() {
        log.info(
            "Checking Tek Module Configuration: [{}]",
            ClassUtils.getUserClass(configuration).getSimpleName()
        );
        checkModuleConfiguration();
        log.info(
            "[{}] success!",
            ClassUtils.getUserClass(configuration).getSimpleName()
        );
    }
}

File: core/src/main/kotlin/com/tek/core/TekModuleConfiguration.kt
package com.tek.core

import com.tek.core.util.LoggerDelegate
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.env.ConfigurableEnvironment
import org.springframework.util.ClassUtils
import javax.annotation.PostConstruct

abstract class TekModuleConfiguration<Configuration>(
    private val configuration: Class<Configuration>
) {

    @Autowired
    lateinit var environment: ConfigurableEnvironment

    companion object {
        private val log by LoggerDelegate()
    }

    @PostConstruct
    fun postConstruct() {
        log.info(
            "Checking Tek Modules Configuration: {}",
            ClassUtils.getUserClass(configuration).simpleName
        )
        checkModuleConfiguration()
    }

    abstract fun checkModuleConfiguration()
}